### PR TITLE
Fix issue 2274 [ Invalidate cache when retrieving register data and executing filtering with params shared across screens]

### LINKED ISSUE #2274: Invalidate cache when retrieving register data and executing filtering with params shared across screens
**Describe the feature request.**
When loading and filtering register data using dataQueries at first they are executed and filtering is working as expected. However, this information is cached in a variable called `configCacheMap` which keeps the track of `Configuration` data. This caches the dataQueries too and when subsequent execution happens, it retrieves the confugration cached with by the `configKey`
The purpose of this ticket is to allow filtering by a dataQuery when defined and value is passed using an `actionParameter`.

**Acceptance criteria**
Filtering by `practitionerId` or any configurable `actionParameter` should work

**How to replicate the issue**
Login to the app using `chss` appId
Define params to filter  with as described [here](https://fhircore.smartregister.org/configurations/user-interface/register-config)
Define `dataQueries` on the target profile or register screen to use the `params` shared above.
Launch `practitioner_profile` and choose any `CHA` then navigate to `household_register`

**Observation**
You will notice filtering is working the first time.
However, when you choose a different `CHA` it still re-uses data from previous `practitionerId`

**Implementation plan (For Engineers)**
To make this configurable when passing data across screens, add config called `isFilterParam` in `ActionParameter data class`
Create a map called `params` in `RegisterRepository` to check whether the `ActionConfigs` passed contains the key above and if it's `true`
Since this only affects register retrieval, pass the `params` and check whether they are null in `ConfigurationRegistry`.
If `params` are empty, use the cached `configuartion` information otherwise retrieve new data.

### PR DESCRIPTION
**Pull request by SebaMutuku**
**IMPORTANT: Where possible all PRs must be linked to a Github issue**

Fixes #2274 

**Engineer Checklist**
- [x] I have written **Unit tests** for any new feature(s) and edge cases for bug fixes
- [x] I have added any strings visible on UI components to the `strings.xml` file
- [x] I have updated the  [CHANGELOG.md](./CHANGELOG.md) file for any notable changes to the codebase
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the project's style guide
- [x] I have built and run the FHIRCore app to verify my change fixes the issue and/or does not break the app 

**Code Reviewer Checklist**
- [ ] I have verified **Unit tests** have been written for any new feature(s) and edge cases
- [ ] I have verified any strings visible on UI components are in the `strings.xml` file
- [ ] I have verifed the [CHANGELOG.md](./CHANGELOG.md) file has any notable changes to the codebase
- [ ] I have verified the solution has been implemented in a configurable and generic way for reuseable components
- [ ] I have built and run the FHIRCore app to verify the change fixes the issue and/or does not break the app

### CODE REVIEW

File: android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/register/RegisterRepository.kt [hunk 1]
```diff
@@ -328,7 +328,9 @@ constructor(
       paramsList
         ?.filter {
           (it.paramType == ActionParameterType.PARAMDATA ||
-            it.paramType == ActionParameterType.UPDATE_DATE_ON_EDIT) && it.value.isNotEmpty()
+            it.paramType == ActionParameterType.UPDATE_DATE_ON_EDIT) &&
+            it.value.isNotEmpty() &&
+            it.isFilterParam == true
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/register/RegisterRepository.kt:333]: Why do we need this new property? can we have a list of all the param types (instead) that at filterable and confirm if the current `paramType` is in the list.

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt [hunk 2]
```diff
@@ -85,7 +85,8 @@ constructor(
   ): T {
     require(!configType.parseAsResource) { "Configuration MUST be a template" }
     val configKey = if (configType.multiConfig && configId != null) configId else configType.name
-    if (configCacheMap.contains(configKey)) return configCacheMap[configKey] as T
+    if (configCacheMap.contains(configKey) && paramsMap?.isEmpty() == true)
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt:88]: Please document why the `paramsMap.isEmpty` check is necessary

**Review by ellykits — CHANGES_REQUESTED** [review]

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt [hunk 3]
```diff
@@ -76,7 +76,10 @@ constructor(
 
   /**
    * Retrieve configuration for the provided [ConfigType]. The JSON retrieved from [configsJsonMap]
-   * can be directly converted to a FHIR resource or hard coded custom model.
+   * can be directly converted to a FHIR resource or hard coded custom model. The filtering assumes
+   * you are passing data across screens and finally used in dataQueries which is later used to
+   * retrieve registerConfiguration. The check for [paramsMap] is empty is therefore necessary to
+   * confirm that actual params used in the dataQuery are passed when retrieving the configurations.
```
Comment (pld) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt:82]:
```suggestion
   * you are passing data across screens, then later using it in DataQueries and to retrieve
   * registerConfiguration. It is necessary to check that [paramsMap] is empty to confirm
   * that the params used in the DataQuery are passed when retrieving the configurations.
```

**Review by pld — APPROVED** [review]

**Review by pld — APPROVED** [review]

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt [hunk 4]
```diff
@@ -85,7 +88,8 @@ constructor(
   ): T {
     require(!configType.parseAsResource) { "Configuration MUST be a template" }
     val configKey = if (configType.multiConfig && configId != null) configId else configType.name
-    if (configCacheMap.contains(configKey)) return configCacheMap[configKey] as T
+    if (configCacheMap.contains(configKey) && paramsMap?.isEmpty() == true)
+      return configCacheMap[configKey] as T
```
Comment (pld) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt:92]: this is a know issue, jacoco can't detect coverage on inline functions

**Review by pld — APPROVED** [review]

**Pull request merged**